src/Enzo/enzo_EnzoEOSIdeal.cpp

```diff
@@ -0,0 +1,409 @@
+	  pressure(iz,iy,ix) = 
+	    gm1 * ((etot(iz,iy,ix) - 0.5 * v2) * density(iz,iy,ix) - 0.5 * b2);
+	}
+
+      }
+    }
+  }
+}
+
+
+void EnzoEOSIdeal::pressure_from_reconstructable
+(Block *block, Grouping &reconstructable_group, std::string pressure_name,
+ int stale_depth, int reconstructed_axis) const
+{
+  // This is necessary since other equations of state may not include pressure
+  // as a reconstructable quantity.
+
+  // We will check if pressure_name the field containing pressure in
+  // reconstructable_group are the same, if so then do nothing. Otherwise,
+  // simply copy the values over.
+
+  if (reconstructable_group.item("pressure",0) == pressure_name){
+    // The fields have the same name, we don't have to do anything
+    return;
+  } else {
+    EnzoFieldArrayFactory array_factory(block, stale_depth);
+    EFlt3DArray old_p = retrieve_field_(array_factory, reconstructable_group,
+					"pressure", 0, reconstructed_axis);
+    
+    Grouping temp_group;
+    temp_group.add(pressure_name, "pressure");
+    EFlt3DArray new_p = retrieve_field_(array_factory, temp_group, "pressure",
+					0, reconstructed_axis);
+
+    for (int iz=0; iz< old_p.shape(0); iz++) {
+      for (int iy=0; iy< old_p.shape(1); iy++) {
+	for (int ix=0; ix< old_p.shape(2); ix++) {
+	  new_p(iz,iy,ix) = old_p(iz,iy,ix);
+	}
+      }
+    }
+    // new_p.subarray() = old_p;
+  }
+}
+
+//----------------------------------------------------------------------
+
+// based on the enzo's hydro_rk implementation of synchronization (found in the
+// Grid_UpdateMHD.C file)
+void EnzoEOSIdeal::apply_floor_to_energy_and_sync(Block *block,
+						  Grouping &integrable_group,
+						  int stale_depth) const
+{
+  if (grackle_variable_gamma_()){ 
+    ERROR("EnzoEOSIdeal::apply_floor_to_energy_and_sync",
+	  "Not equipped to handle grackle and spatially variable gamma");
+  }
+
+  const bool idual = this->uses_dual_energy_formalism();
+  // in hydro_rk, eta was set equal to eta1 (it didn't use eta2 at all)
+  const double eta = dual_energy_formalism_eta_;
+
+  EnzoFieldArrayFactory array_factory(block, stale_depth);
+  EFlt3DArray density, vx, vy, vz, etot, eint, bx, by, bz;
+
+  // We are going to check that these are the specified fields
+  density = array_factory.from_grouping(integrable_group, "density", 0);
+  vx = array_factory.from_grouping(integrable_group, "velocity", 0);
+  vy = array_factory.from_grouping(integrable_group, "velocity", 1);
+  vz = array_factory.from_grouping(integrable_group, "velocity", 2);
+  etot = array_factory.from_grouping(integrable_group, "total_energy", 0);
+  if (idual){
+    eint = array_factory.from_grouping(integrable_group, "internal_energy", 0);
+  }
+  bx = array_factory.from_grouping(integrable_group, "bfield", 0);
+  by = array_factory.from_grouping(integrable_group, "bfield", 1);
+  bz = array_factory.from_grouping(integrable_group, "bfield", 2);
+
+  float ggm1 = get_gamma()*(get_gamma() - 1.);
```

Comment:
Yeah, but I don't think we need to worry about this right now. It will only become expensive if there are spatial variations, and we will need to change this code anyway then.